src/pyhf/infer/calculators.py

```diff
@@ -0,0 +1,155 @@
+    bestfit_nuisance_asimov = fixed_poi_fit(asimov_mu, data, pdf, init_pars, par_bounds)
+    return pdf.expected_data(bestfit_nuisance_asimov)
+
+
+class AsymptoticTestStatDistribution(object):
+    """
+    The distribution the test statistic in the asymptotic case.
+
+    Note: These distributions are in -µ^/sigma space. In the ROOT
+    implementation the same sigma is assumed for both hypotheses
+    and p-values etc are computed in that space. This assumption
+    is necessarily valid, but we keep this for compatibility
+    reasons.
+
+    In the -µ^/sigma space, the test statistic (i.e. µ^/sigma) is
+    normally distributed with unit variance and its mean at 
+    the -µ', where µ' is the true poi value of the hypothesis.
+    """
+
+    def __init__(self, shift):
+        """
+        Asymptotic test statistic distribution.
+
+        Args:
+            shift: the displacement of the test statistic distribution
+
+        Returns:
+            distribution
+        
+        """
+        self.shift = shift
+
+    def pvalue(self, value):
+        """
+        Compute the p-value for a given value of the test statistic.
+        
+        Args:
+            value: the test statistic value.
+
+        Returns;
+            pvalue (float): the integrated probability to observe
+            a value at least as large as the observed one.
+
+        """
+        tensorlib, _ = get_backend()
+        return 1 - tensorlib.normal_cdf(value - self.shift)
+
+    def expected_value(self, nsigma):
+        """
+        Return the expected value of the test statistic.
+
+        Args:
+            nsigma: number of standard deviations.
+
+        Returns;
+            expected value (float): the expected value of the test statistic.
+        
+        """
+        return nsigma
+
+
+class AsymptoticCalculator(object):
+    """The Asymptotic Calculator."""
+
+    def __init__(self, data, pdf, init_pars=None, par_bounds=None, qtilde=False):
+        """
+        Asymptotic Calculator.
+
+        Args:
+            data: the observed data
+            pdf: the statistical model
+            init_pars: the initial parameters to be used for fitting
+            par_bounds: the parameter bounds used for fitting
+
+        Returns:
+            calculator
+             
+        """
+        self.data = data
+        self.pdf = pdf
+        self.init_pars = init_pars or pdf.config.suggested_init()
+        self.par_bounds = par_bounds or pdf.config.suggested_bounds()
+        self.qtilde = qtilde
+
+    def distributions(self, poi_test):
+        """
+        Probability Distributions of the test statistic value under the signal + background and and background-only hypothesis.
+
+        Args:
+            poi_test: the value for the parameter of interest.
+
+        Returns
+            distributions (Tuple of `AsymptoticTestStatDistribution`): the distributions under the hypotheses.
```

Comment:
Could link directly in the docs:

```suggestion
            distributions (Tuple of ~pyhf.infer.calculators.AsymptoticTestStatDistribution): the distributions under the hypotheses.
```